### PR TITLE
Fixed #24042 -- Added db.models.Field.db_generated

### PR DESCRIPTION
**Pull request by lucaswiman**
References https://code.djangoproject.com/ticket/24042

[The documentation](​https://docs.djangoproject.com/en/dev/howto/custom-model-fields/#emulating-built-in-field-types) on `Field.get_internal_type` implies that it should define custom behavior and ensure that database-correct behavior is followed. For Postgres (though not for sqlite), specifying `get_internal_type()` to be "AutoField" fails to correctly set the field, leading to errors like this:

```pytb
test_auto_field (field_subclassing.tests.CustomField) ... ERROR

======================================================================
ERROR: test_auto_field (field_subclassing.tests.CustomField)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/lucaswiman/opensource/django/tests/field_subclassing/tests.py", line 128, in test_auto_field
    o.save()
  File "/Users/lucaswiman/opensource/django/django/db/models/base.py", line 694, in save
    force_update=force_update, update_fields=update_fields)
  File "/Users/lucaswiman/opensource/django/django/db/models/base.py", line 722, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/Users/lucaswiman/opensource/django/django/db/models/base.py", line 803, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/Users/lucaswiman/opensource/django/django/db/models/base.py", line 842, in _do_insert
    using=using, raw=raw)
  File "/Users/lucaswiman/opensource/django/django/db/models/manager.py", line 131, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/lucaswiman/opensource/django/django/db/models/query.py", line 954, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/Users/lucaswiman/opensource/django/django/db/models/sql/compiler.py", line 931, in execute_sql
    cursor.execute(sql, params)
  File "/Users/lucaswiman/opensource/django/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/Users/lucaswiman/opensource/django/django/db/utils.py", line 95, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/Users/lucaswiman/opensource/django/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
IntegrityError: null value in column "id" violates not-null constraint
DETAIL:  Failing row contains (null).
```
The attached diff adds a regression test which fails on postgres databases with the error above. I also fixed all other uses of `isinstance(..., AutoField)` I could find in the code.

The immediate motivation for this is writing integration tests which use django-salesforce on a CI system that uses a postgres database. I've filed a separate issue there, though I think fixing the behavior in Django to match the documentation would be a preferable solution. [The package maintainer for django-salesforce agrees](​https://github.com/django-salesforce/django-salesforce/issues/90#issuecomment-67932443).

### How to Reproduce

Check out the master branch and apply my diff from the tests/ directory. Run the tests with a settings file pointing to a postgres database:

```
PYTHONPATH=..:$PYTHONPATH ./runtests.py --verbosity=2 --failfast field_subclassing --settings=test_postgres
```
It will fail with the above stack trace.

### Testing

I verified that the regression test fails without the remainder of the diff. I also verified that the tests still pass on the default test_sqlite file.

### CODE REVIEW

**Comment by lucaswiman** [comment]
@mjtamlyn @akaariai @timgraham sorry for the delay in updating the implementation of the patch. I've made the following changes:
- Added a `db_generated` field to `models.Field` (where it is false), and to `models.AutoField` (where it is true).
- Changed various places in the code which depend on `_meta.has_auto_field` to use `_meta.pk.db_generated`.
- Removed `InlineAdminForm.field_count`.  The function seems to be both broken and unused, since it references a missing `InlineAdminForm.has_auto_field()` method, and is not in fact used in `tabular.html` as claimed in the comment.

I verified that the `tests.field_subclassing` tests pass using both Postgres and sqlite, and that the full suite of tests has the same collection of failures between the master branch and this branch. The [test failures](https://gist.github.com/lucaswiman/dd8cddd0be68f58b89dc) I saw appear unrelated.

**Comment by charettes** [comment]
Buildbot, test this please.

**Comment by lucaswiman** [comment]
Looking at the build failures for mysql and postgis.

**Comment by lucaswiman** [comment]
@charettes, the Python2 MySQL and PostGIS failures should be fixed by 9e0081c9fa762ac1b1ee06a8b58ce85e5b49b8f1.

**Comment by charettes** [comment]
Buildbot, test this please.

**Comment by lucaswiman** [comment]
Looks like a green build. Let me know if there are other changes that are needed, or where you'd suggest updating the docs to reflect this change. Thanks!

**Comment by timgraham** [comment]
Docs would go at least in docs/ref/models/fields.txt and releases/1.9.txt. See our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) for how to get your patch in a good state. Be sure to update the flags on the ticket so it appears in the review queue.

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -127,6 +127,10 @@ class Field(RegisterLookupMixin):
     one_to_one = None
     related_model = None
 
+    # This field tracks whether the value of the field is generated by the
+    # database backend.
+    db_generated = False
```
Comment (akaariai) [inline django/db/models/fields/__init__.py]: DB generated works for primary keys only, we should mention it here.

**Comment by timgraham** [comment]
Closing due to inactivity. Please send a new PR if you can continue work on this, thanks!

**Comment by owais** [comment]
@timgraham I'm interested in this and would like to help get this done. It seems only thing pending is updating the docs and bringing the PR up to date with the current state of the codebase. Is there something else that I should look at before taking this on?

**Comment by timgraham** [comment]
The [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) is a good place to start if you're a new contributor. Thanks!

**Comment by owais** [comment]
Thanks

**Comment by lucaswiman** [comment]
@owais Thanks for taking this up!

This change is probably worth making, so let me know if you have questions about it. I stopped working on it mostly because I felt like I'd gotten out of my comfort zone for an initial contribution to Django, and the direct reason I was working on it was obviated by https://github.com/django-salesforce/django-salesforce/issues/90. I'm not very familiar with Django's contribution procedures (which I thought I'd followed, but failed to), and the scope kept increasing (from fixing a minor inconsistency with the docs for `get_internal_type` to more delicate refactoring/changes in the ORM).

**Pull request closed without merging**